[TEST] Implement HotThreads unit tests

grcevski opened the pull request:
Add unit tests for:

- The internal HotThreads logic for calculating and sorting threads by cpu, wait and blocked "hotness".
- Tests for identifying certain threads as idle. 
- Tests for supported report types (e.g. cpu, wait, blocked).

This test PR precedes any changes and refactoring to make HotThreads more accurate.

Commit:
[TEST] Implement HotThreads unit tests

Add unit tests for the internal HotThreads logic for calculating and
sorting threads by CPU, Wait and Blocked "hotness". Also adds tests
for identifying certain threads as idle, as well as supported report
types (e.g. cpu, wait, blocked).

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

dnhatn commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/hotthreads_unit_tests

dnhatn commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -0,0 +1,291 @@
+            assertIdleThreadHelper(idleThread, bottomOfStack);
+
+            var middleOfStack = new ArrayList<>(testJvmStack);
+            middleOfStack.add(1, extraFrame);
```
Maybe use a random position: `middleOfStack.add(between(0, testJvmStack.size()), extraFrame);`?

grcevski replied:
Good idea! I'll set it up such that it's random between start + 1, end-1 so that I don't accidentally duplicate what I'm already testing with top and bottom.

dnhatn commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -0,0 +1,291 @@
+        when(threadOne.getThreadName()).thenReturn("Thread One");
+        when(threadOne.getStackTrace()).thenReturn(testCases[0].one);
+
+        assertEquals(hotThreads.similarity(threadOne, null), 0);
```
Same here, `0` should be the expectation?

grcevski replied:
👍

dnhatn commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -0,0 +1,291 @@
+            when(threadTwo.getThreadName()).thenReturn("Thread Two");
+            when(threadTwo.getStackTrace()).thenReturn(testCase.two);
+
+            assertEquals(hotThreads.similarity(threadOne, threadTwo), testCase.similarityScore);
```
`testCase.testCase.similarityScore` should be the expectation?

grcevski replied:
Good catch! Thanks!

dnhatn commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -0,0 +1,291 @@
+        ThreadMXBean mockedMXBean = mock(ThreadMXBean.class);
+        when(mockedMXBean.isThreadCpuTimeSupported()).thenReturn(true);
+
+        // Using negative thread ids to ensure they don't collide with our JUnit thread id.
```
I don't understand this comment as we are mocking stuff now. Can you explain it?

grcevski replied:
I see how my comment is unclear now :). I'll expand the comment with a better description. 

Essentially, the code inside innerDetect calls Thread.currentThread().getThreadId(), to exclude the HotThreads runner thread itself from the report. While it's highly unlikely that the Thread.currentThread().getThreadId() will return a number between [1, 4], it might happen and the test will become flaky, effectively it will fail to report one of the ids that matches the threadId of the current JUnit thread.

I could eliminate this issue by mocking Thread.currentThread(), however it's a static call and I'd need PowerMockito to mock it.

But... now that I think about it, the code only needs the current thread id and not really the current Thread reference. So with a small refactor to the caller arguments of innerDetect I can eliminate the static call, and supply non-matching thread id for the test. This will simplify all those -/+1 conversions too. This will be better, I'll go with the latter approach.

Thanks!

dnhatn replied:
Thank you for the explanation. I am fine with the current code.

grcevski replied:
Your question was spot on, it got me thinking more about it and with the small change to pass the threadId I was able to add one more test. This new test ensures that the thread processing code does indeed skip the HotThreads request thread itself.

dnhatn reviewed (commented):
I left some minor comments, but the new tests look great. Thanks @grcevski!

Commit:
Refactor test code based on PR comments

Implemented random middle of the stack position, more solid current
thread avoidance, fixed assertEquals expected and value argument
ordering.

Commit:
Add test to check if the HotThreads skips itself

Added test to ensure the current thread running the HotThreads request
is not in the list of reported threads, even when it's the only thread
hogging the CPU.

dnhatn reviewed (approved):
LGTM. Thanks @grcevski!

You can add the label `auto-backport` to ask the bot to backport this change to 7.x.

grcevski commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -0,0 +1,317 @@
+            assertIdleThreadHelper(idleThread, bottomOfStack);
+
+            var middleOfStack = new ArrayList<>(testJvmStack);
+            middleOfStack.add(between(Math.min(1, testJvmStack.size()), Math.max(0, testJvmStack.size()-1)), extraFrame);
```
Ah! Good point. I'll add a check for size > 1, if it's <=1 the previous cases cover the positioning.

Commit:
Small test fix for between usage

Ensure test code doesn't break with modifying testcase input data
to contain a stack trace of size 1 or less.

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  7.x  | Commit could not be cherrypicked due to conflicts |

To backport manually run `backport --upstream elastic/elasticsearch --pr 76857`
